### PR TITLE
Handling of flag(ZVOL_OP_FLAG_READ_METADATA) for reading metadata for read command

### PR DESCRIPTION
**Pull request by mynktl**
Handling of ZVOL_OP_FLAG_READ_METADATA  flag from target for reading metadata.
With this change, ZREPL will read metadata in two cases : 
1. If the read command is meant for the rebuild
2. If read command has ZVOL_OP_FLAG_READ_METADATA set in the zvol_io_hdr flag.

Signed-off-by: mayank <mayank.patel@cloudbyte.com>

### CODE REVIEW

File: include/zrepl_prot.h [hunk 1]
```diff
@@ -50,7 +50,9 @@ extern "C" {
 #define	MAX_IP_LEN	64
 #define	TARGET_PORT	6060
 
-#define	ZVOL_OP_FLAG_REBUILD 0x01
+#define	ZVOL_OP_FLAG_NONE		0x00
```
Comment (jkryl) [inline include/zrepl_prot.h]: I find this unusual. If a field is a set of bits (flags) then it goes without saying that the zero is the default value. This can lead to confusion when default - zero value is regarded as if it was a real flag value.

File: tests/cbtest/gtest/test_zrepl_prot.cc [hunk 2]
```diff
@@ -1020,7 +1020,52 @@ TEST_F(ZreplDataTest, RebuildFlag) {
...
+	struct zvol_io_rw_hdr write_hdr;
+	struct zrepl_status_ack status;
+	struct mgmt_ack mgmt_ack;
+	char buf[4096];
+	int rc;
+
+	/* write a data block with known ionum */
+	write_data_and_verify_resp(m_data_fd1, m_ioseq1, 0, 654);
+
+	sleep(5);
```
Comment (jkryl) [inline tests/cbtest/gtest/test_zrepl_prot.cc]: we should avoid sleeps with magic constants in our tests which are inherently unreliable. If it is necessary then we need some comment and explanation why is that here.

File: tests/cbtest/gtest/test_zrepl_prot.cc [hunk 3]
```diff
@@ -1020,7 +1020,52 @@ TEST_F(ZreplDataTest, RebuildFlag) {
...
+	ASSERT_EQ(rc, sizeof (read_hdr));
+	ASSERT_EQ(read_hdr.io_num, 654);
+	ASSERT_EQ(read_hdr.len, sizeof (buf));
+	rc = read(m_data_fd1, buf, sizeof (buf));
+	ASSERT_ERRNO("read", rc >= 0);
+	ASSERT_EQ(rc, sizeof (buf));
+}
+
+/*
+ * Metadata ionum should be returned when equested by
```
Comment (jkryl) [inline tests/cbtest/gtest/test_zrepl_prot.cc]: typo equested -> requested.

File: tests/cbtest/gtest/test_zrepl_prot.cc (hunk 2 above)
Comment (mynktl) [reply tests/cbtest/gtest/test_zrepl_prot.cc]: ok. I think we can avoid sleep here. I will remove sleep.

File: tests/cbtest/gtest/test_zrepl_prot.cc (hunk 3 above)
Comment (mynktl) [reply tests/cbtest/gtest/test_zrepl_prot.cc]: ok

File: include/zrepl_prot.h (hunk 1 above)
Comment (mynktl) [reply include/zrepl_prot.h]: Ok. I agree with you. I will remove this.

**Comment by mynktl** [comment]
@jkryl  I have done changes as per review comments. Please have a look and let me know if more changes needed.

**Review by jkryl — APPROVED** [review]
thank you!

**Review by gila — APPROVED** [review]
